api/v3/PaymentProcessor.php

```diff
@@ -115,3 +115,44 @@ function _civicrm_api3_payment_processor_getlist_defaults(&$request) {
     ),
   );
 }
+
+/**
+ * Action refund.
+ *
+ * @param array $params
+ *
+ * @return array
+ *   API result array.
+ * @throws CiviCRM_API3_Exception
+ */
+function civicrm_api3_payment_processor_refund($params) {
+  $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
+  $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', array('id' => $params['payment_processor_id'])));
+  if (!$processor->supportsRefund()) {
+    throw API_Exception('Payment Processor does not support refund');
+  }
+  $result = $processor->doRefund($params);
+  if (is_a($result, 'CRM_Core_Error')) {
+    throw API_Exception('Payment failed');
+  }
+  return civicrm_api3_create_success(array($result), $params);
+}
+
+/**
+ * Action Refund.
+ *
+ * @param array $params
+ *
+ */
+function _civicrm_api3_payment_processor_refund_spec(&$params) {
+  $params['payment_processor_id']['api.required'] = 1;
+  $params['trxn_id'] = [
```

Comment:
Hmm - so this is the question as to what different processors will require to do the refund - I'm not 100% sure - some might just need 'contact_id' or payment_token_id - I'd probably leave this out until we have worked through a few

Reply:
Agree that params may well be different.  I've seen different processors use trxn_id or processor_id for example (all my processors use trxn_id, don't get me started on the internal `subscriptionId` (which normally maps to trxn_id) used by various parts of the payment system :-)  But obviously we do need a standard way to identify the actual payment we want to refund.

Reply:
As per my research, I found that trxn_id (of last payment made) is a necessary field for claiming a refund. Like for Moneris trxn_id (as trxn_number) and invoice_id (as order_id) is a necessary field send a refund request (I present some example links). As I mentioned above along with trxn_id there are other parameters which might vary on PP basis. But for the time being, we can remove the required property and let PP can throw the respective exceptions on their end. What do you think

Reply:
Based on @monishdeb comments I do think we are better specifying *some* mandatory parameters at this stage and if necessary removing them later if we discover they are not universal.  As this will help enforce more consistent code on the payment processor.